File: hooks/stop-gate.py
#!/usr/bin/env python3
"""stop hook: workflow stop gate.

Replaces the older `completion-summary-audit.py` with a brief, lifecycle-style
name. Behavior is a strict superset:

- always emit a final-verification reminder;
- if the stop event reports an error and we are still inside the configured
  loop limit, allow exactly one conservative follow-up; and
- if a workflow-state document is reachable (via `OH_MY_CURSOR_WORKFLOW_STATE`
  env var or a `workflow_state` field in the stop payload), surface a concise
  summary of pending acceptance criteria so the user can finish them before
  closing the session.

The hook is fail-open: any malformed input or unreadable state file falls back
to the generic reminder instead of blocking the user.
"""
from __future__ import annotations

import json
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))
from _trace import trace as _trace  # noqa: E402

# Resolve ROOT from environment or workspace, not __file__ (which may be unreliable in hook contexts)
ROOT = Path(os.environ.get("OH_MY_CURSOR_WORKSPACE", os.getcwd())).resolve()
if not (ROOT / "hooks").exists() or "plugins/local/oh-my-cursor" in str(ROOT):
    ROOT = Path(__file__).resolve().parents[1]
ERROR_STATUSES = {"error", "failed", "failure"}


def _read_payload() -> dict:
    raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        value = json.loads(raw)
    except json.JSONDecodeError:
        return {"_invalid_json": True}
    return value if isinstance(value, dict) else {"payload": value}


def _first_string(payload: dict, keys: tuple[str, ...]) -> str:
    for key in keys:
        value = payload.get(key)
        if isinstance(value, str):
            return value
    return ""


def _loop_count(payload: dict) -> int:
    for key in ("loop_count", "loopCount", "current_loop", "currentLoop"):
        value = payload.get(key)
        if isinstance(value, int):
            return value
        if isinstance(value, str) and value.isdigit():
            return int(value)
    return 0


def _resolve_state_path(payload: dict) -> Path | None:
    """Locate the workflow-state document for this stop event.

    Resolution order:
    1. ``OH_MY_CURSOR_WORKFLOW_STATE`` environment variable -- override-only,
       intended for unusual setups (per-task archives under
       ``docs/plans/<task-id>/``, alternative workspace layouts, tests).
    2. ``workflow_state`` field on the stop payload -- same override semantics
       provided per-event by the host.
    3. The canonical default ``.cursor/state/workflow-state.json`` (matches
       the path the cursor-state-bridge writes when no ``task_id`` is set,
       and the path ``state-watcher.py`` validates after every bridge write).
    """
    candidates: list[str] = []
    env_path = os.environ.get("OH_MY_CURSOR_WORKFLOW_STATE")
    if env_path:
        candidates.append(env_path)
    payload_path = payload.get("workflow_state") if isinstance(payload, dict) else None
    if isinstance(payload_path, str):
        candidates.append(payload_path)
    candidates.append(str(ROOT / ".cursor" / "state" / "workflow-state.json"))
    for raw in candidates:
        try:
            candidate = Path(raw).expanduser()
            if not candidate.is_absolute():
                candidate = (ROOT / candidate).resolve()
            if candidate.is_file():
                return candidate
        except (OSError, RuntimeError, ValueError):
            continue
    return None


def _summarize_state(state_path: Path) -> dict:
    try:
        data = json.loads(state_path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError):
        return {"loaded": False}

    criteria = data.get("acceptance_criteria")
    pending: list[str] = []
    failed: list[str] = []
    if isinstance(criteria, list):
        for item in criteria:
            if not isinstance(item, dict):
                continue
            status = str(item.get("status", "")).lower()
            label = str(item.get("id") or item.get("criterion") or "").strip()
            if not label:
                continue
            if status in {"failed"}:
                failed.append(label)
            elif status not in {"passed", "skipped"}:
                pending.append(label)

    return {
        "loaded": True,
        "task_id": data.get("task_id"),
        "phase": data.get("phase"),
        "status": data.get("status"),
        "current_role": data.get("current_role"),
        "next_action": data.get("next_action"),
        "pending_criteria": pending[:10],
        "failed_criteria": failed[:10],
    }


def main() -> int:
    payload = _read_payload()
    if payload.get("_invalid_json"):
        print(json.dumps({"status": "pass", "continue": False, "message": "Stop hook input was not JSON; skipped audit."}))
        return 0

    status = _first_string(payload, ("status", "final_status", "outcome", "result")).lower()
    loop_count = _loop_count(payload)
    should_continue = status in ERROR_STATUSES and loop_count < 1

    base_message = (
        "Before final delivery, verify every acceptance criterion with fresh evidence and keep runtime claims bounded to checked-in artifacts plus actual smoke results."
    )
    if should_continue:
        base_message = (
            "The stop event reports an error. One conservative follow-up is allowed to collect failure evidence, fix only the blocking issue, and rerun the relevant check."
        )

    state_path = _resolve_state_path(payload)
    state_summary = _summarize_state(state_path) if state_path else {"loaded": False}

    parts = [base_message]
    if state_summary.get("loaded"):
        if state_summary.get("failed_criteria"):
            parts.append(
                "Workflow state reports failed acceptance criteria: "
                + ", ".join(state_summary["failed_criteria"])
                + "."
            )
        if state_summary.get("pending_criteria"):
            parts.append(
                "Workflow state still has pending acceptance criteria: "
                + ", ".join(state_summary["pending_criteria"])
                + "."
            )
        if state_summary.get("next_action"):
            parts.append(f"Next recorded action: {state_summary['next_action']}.")

    user_message = " ".join(parts)

    output = {
        "status": "followup-requested" if should_continue else "pass",
        "continue": should_continue,
        "loop_limit": 1,
        "loop_count": loop_count,
        "user_facing_message": user_message,
        "additional_context": user_message if should_continue else "",
        "workflow_state": {
            "path": state_path.as_posix() if state_path else None,
            "loaded": state_summary.get("loaded", False),
            "task_id": state_summary.get("task_id"),
            "phase": state_summary.get("phase"),
            "status": state_summary.get("status"),
            "current_role": state_summary.get("current_role"),
            "pending_criteria": state_summary.get("pending_criteria", []),
            "failed_criteria": state_summary.get("failed_criteria", []),
        },
    }
    _trace({
        "hook": "stop-gate",
        "event": "stop",
        "status": output.get("status"),
        "continue": should_continue,
        "loop_count": loop_count,
        "state_loaded": state_summary.get("loaded", False),
        "pending_criteria": state_summary.get("pending_criteria", [])[:5],
        "failed_criteria": state_summary.get("failed_criteria", [])[:5],
    })
    print(json.dumps(output, ensure_ascii=False))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: hooks/stop-gate.ts
import * as fs from 'node:fs';
import * as path from 'node:path';
import * as process from 'node:process';
import { fileURLToPath } from 'node:url';
import { resolveWorkspaceRoot } from './_repo.ts';
import { trace } from './_trace.ts';

const currentFile = fileURLToPath(import.meta.url);
const WORKSPACE_ROOT = resolveWorkspaceRoot(currentFile);

const ERROR_STATUSES = new Set(['error', 'failed', 'failure']);

function readPayload(): any {
  try {
    const raw = fs.readFileSync(0, 'utf-8');
    if (!raw.trim()) {
      return {};
    }
    return JSON.parse(raw);
  } catch (err: any) {
    return { _invalid_json: true };
  }
}

function firstString(payload: any, keys: string[]): string {
  if (!payload || typeof payload !== 'object') return '';
  for (const key of keys) {
    const value = payload[key];
    if (typeof value === 'string') {
      return value;
    }
  }
  return '';
}

function getLoopCount(payload: any): number {
  if (!payload || typeof payload !== 'object') return 0;
  for (const key of ['loop_count', 'loopCount', 'current_loop', 'currentLoop']) {
    const value = payload[key];
    if (typeof value === 'number') {
      return value;
    }
    if (typeof value === 'string' && /^\d+$/.test(value)) {
      return parseInt(value, 10);
    }
  }
  return 0;
}

function resolveStatePath(payload: any): string | null {
  const candidates: string[] = [];
  const envPath = process.env.OH_MY_CURSOR_WORKFLOW_STATE;
  if (envPath) {
    candidates.push(envPath);
  }
  const payloadPath = payload?.workflow_state;
  if (typeof payloadPath === 'string') {
    candidates.push(payloadPath);
  }
  candidates.push(path.join(WORKSPACE_ROOT, '.cursor', 'state', 'workflow-state.json'));

  for (const raw of candidates) {
    try {
      let candidate = raw;
      if (candidate.startsWith('~')) {
        const homedir = process.env.HOME || '';
        candidate = path.join(homedir, candidate.slice(1));
      }
      if (!path.isAbsolute(candidate)) {
        candidate = path.resolve(WORKSPACE_ROOT, candidate);
      }
      if (fs.existsSync(candidate) && fs.statSync(candidate).isFile()) {
        return candidate;
      }
    } catch {
      continue;
    }
  }
  return null;
}

function summarizeState(statePath: string): any {
  try {
    const data = JSON.parse(fs.readFileSync(statePath, 'utf-8'));
    if (!data || typeof data !== 'object') {
      return { loaded: false };
    }
    const criteria = data.acceptance_criteria;
    const pending: string[] = [];
    const failed: string[] = [];
    if (Array.isArray(criteria)) {
      for (const item of criteria) {
        if (!item || typeof item !== 'object') continue;
        const status = String(item.status || '').toLowerCase();
        const label = String(item.id || item.criterion || '').trim();
        if (!label) continue;
        if (status === 'failed') {
          failed.push(label);
        } else if (status !== 'passed' && status !== 'skipped') {
          pending.push(label);
        }
      }
    }
    return {
      loaded: true,
      task_id: data.task_id,
      phase: data.phase,
      status: data.status,
      current_role: data.current_role,
      next_action: data.next_action,
      pending_criteria: pending.slice(0, 10),
      failed_criteria: failed.slice(0, 10),
    };
  } catch {
    return { loaded: false };
  }
}

function main(): number {
  const payload = readPayload();
  if (payload?._invalid_json) {
    console.log(JSON.stringify({
      status: 'pass',
      continue: false,
      message: 'Stop hook input was not JSON; skipped audit.'
    }));
    return 0;
  }

  const status = firstString(payload, ['status', 'final_status', 'outcome', 'result']).toLowerCase();
  const loopCount = getLoopCount(payload);
  const shouldContinue = ERROR_STATUSES.has(status) && loopCount < 1;

  let baseMessage = (
    'Before final delivery, verify every acceptance criterion with fresh evidence and keep runtime claims bounded to checked-in artifacts plus actual smoke results.'
  );
  if (shouldContinue) {
    baseMessage = (
      'The stop event reports an error. One conservative follow-up is allowed to collect failure evidence, fix only the blocking issue, and rerun the relevant check.'
    );
  }

  const statePath = resolveStatePath(payload);
  const stateSummary = statePath ? summarizeState(statePath) : { loaded: false };

  const parts = [baseMessage];
  if (stateSummary.loaded) {
    if (stateSummary.failed_criteria && stateSummary.failed_criteria.length > 0) {
      parts.push(`Workflow state reports failed acceptance criteria: ${stateSummary.failed_criteria.join(', ')}.`);
    }
    if (stateSummary.pending_criteria && stateSummary.pending_criteria.length > 0) {
      parts.push(`Workflow state still has pending acceptance criteria: ${stateSummary.pending_criteria.join(', ')}.`);
    }
    if (stateSummary.next_action) {
      parts.push(`Next recorded action: ${stateSummary.next_action}.`);
    }
  }

  const userMessage = parts.join(' ');

  const output = {
    status: shouldContinue ? 'followup-requested' : 'pass',
    continue: shouldContinue,
    loop_limit: 1,
    loop_count: loopCount,
    user_facing_message: userMessage,
    additional_context: shouldContinue ? userMessage : '',
    workflow_state: {
      path: statePath,
      loaded: stateSummary.loaded,
      task_id: stateSummary.task_id || null,
      phase: stateSummary.phase || null,
      status: stateSummary.status || null,
      current_role: stateSummary.current_role || null,
      pending_criteria: stateSummary.pending_criteria || [],
      failed_criteria: stateSummary.failed_criteria || [],
    },
  };

  trace({
    hook: 'stop-gate',
    event: 'stop',
    status: output.status,
    continue: shouldContinue,
    loop_count: loopCount,
    state_loaded: stateSummary.loaded,
    pending_criteria: (stateSummary.pending_criteria || []).slice(0, 5),
    failed_criteria: (stateSummary.failed_criteria || []).slice(0, 5),
  });

  console.log(JSON.stringify(output));
  return 0;
}

process.exit(main());
